src/mame/drivers/gi6809.cpp

```diff
@@ -0,0 +1,1039 @@
+*********************************************/
+
+void gi6809_state::gi6809_videoram_w(offs_t offset, uint8_t data)
+{
+	m_videoram[offset] = data;
+	m_bg_tilemap->mark_tile_dirty(offset);
+}
+
+void gi6809_state::gi6809_colorram_w(offs_t offset, uint8_t data)
+{
+	m_colorram[offset] = data;
+	m_bg_tilemap->mark_tile_dirty(offset);
+}
+
+
+TILE_GET_INFO_MEMBER(gi6809_state::get_bg_tile_info)
+{
+/*  - bits -
+    7654 3210
+    --xx xx--   tiles color.
+    ---- --x-   tiles bank.
+    ---- ---x   tiles extended address (MSB).
+    xx-- ----   unused.
+*/
+
+	int attr = m_colorram[tile_index];
+	int code = ((attr & 1) << 8) | m_videoram[tile_index];
+	int bank = (attr & 0x02) >> 1;      // bit 1 switch the gfx banks
+	int color = (attr & 0x3c) >> 2;     // bits 2-3-4-5 for color
+
+	tileinfo.set(bank, code, color, 0);
+}
+
+void gi6809_state::video_start()
+{
+	m_bg_tilemap = &machine().tilemap().create(*m_gfxdecode, tilemap_get_info_delegate(*this, FUNC(gi6809_state::get_bg_tile_info)), TILEMAP_SCAN_ROWS, 8, 8, 32, 32);
+}
+
+uint32_t gi6809_state::screen_update_gi6809(screen_device &screen, bitmap_rgb32 &bitmap, const rectangle &cliprect)
+{
+	m_bg_tilemap->draw(screen, bitmap, cliprect, 0, 0);
+	return 0;
+}
+
+
+void gi6809_state::gi6809_palette(palette_device &palette) const
+{
+/*
+    This hardware has a feature called BLUE KILLER.
+    Using the original intensity line, the PCB has a bridge
+    that allows (as default) to turn the background black.
+
+    All games running on this hardware
+    were designed with black background.
+
+    7654 3210
+    ---- ---x   red component.
+    ---- --x-   green component.
+    ---- -x--   blue component.
+    ---- x---   blue killer.
+    xxxx ----   unused.
+*/
+
+	// 0000KBGR
+	uint8_t const *const color_prom = memregion("proms")->base();
+	if (!color_prom)
+		return;
+
+	for (int i = 0; i < palette.entries(); i++)
+	{
+		// blue killer (from schematics)
+		int const bk = BIT(color_prom[i], 3);
+
+		// red component
+		int const r = BIT(color_prom[i], 0) * 0xff;
+
+		// green component
+		int const g = BIT(color_prom[i], 1) * 0xff;
+
+		// blue component
+		int const b = bk * BIT(color_prom[i], 2) * 0xff;
+
+		palette.set_pen_color(i, rgb_t(r, g, b));
+	}
+}
+
+
+/*******************************************
+*               Read Handlers              *
+*******************************************/
+
+uint8_t gi6809_state::gi6809_mux_port_r()
+{
+	uint8_t data = 0xff;
+
+	for (int i = 0; i < 4 ; i++)
+		if (BIT(~m_mux_data, i + 4))
+		{
+			data &= m_input[i]->read();
+			if(i == 1)
+			{
+				if((m_coin_gl->read() == 0) || (m_stage != 0))
+				{
+					switch(m_stage)
+					{
+						case 0: m_stage++; return data & 0xfb;  // upper coin
+						case 1: m_stage++; return data & 0xeb;  // upper coin + lower coin
+						case 2: m_stage++; return data & 0xef;  // lower coin
+						case 3: m_stage++; return data & 0xff;  // none
+						case 4: m_stage = 0; break;
+					}
+				}
+			}
+		}
```

Comment:
Besides the inconsistent spacing, are you really saying it has a two-way multiplexer controlled by bit 5 of the matrix drive?  If it does, you should lift the part for handling that out of the loop.  But that doesn’t make a lot of sense anyway.  It’s more likely the mod 5 counter just enables reading one of the bits or something.  The code as it is now rings alarm bells.